File: python/tests/test_hooks.py
"""
python/tests/test_hooks.py — smoke-test pybind11 wrappers for the
extension hooks (PnLTracker, StorageSink, RiskManager, KillSwitch,
OrderValidator, MarketDataRecorderHook, ReplaySource, Executor,
ExecutionListener, set_log_callback).

Run from repo root:
    PYTHONPATH=build/python python3 python/tests/test_hooks.py
"""

import os
import sys

build_dir = os.path.join(os.path.dirname(__file__), "..", "..", "build", "python")
sys.path.insert(0, os.path.abspath(build_dir))

import flox_py as flox

_passed = 0
_failed = 0


def check(cond, msg):
    global _passed, _failed
    if cond:
        print(f"  ok  {msg}")
        _passed += 1
    else:
        print(f"  FAIL  {msg}")
        _failed += 1


def make_runner_with_strategy():
    reg = flox.SymbolRegistry()
    sym_id = reg.add_symbol("test", "BTC", 0.01)

    fired_signal = []

    def on_signal(sig):
        fired_signal.append(sig)

    runner = flox.Runner(reg, on_signal, threaded=False)

    # Minimal strategy that fires a market buy on first trade.
    fired = [False]

    class S(flox.Strategy):
        def on_trade(self, ctx, trade):
            if fired[0]:
                return
            fired[0] = True
            self.emit_market_buy(sym_id, 1.0)

    strat = S(symbols=[sym_id])
    runner.add_strategy(strat)
    return reg, runner, strat, sym_id, fired_signal


def test_pnl_tracker():
    print("test_pnl_tracker")
    reg, runner, strat, sym_id, _ = make_runner_with_strategy()

    seen = []

    class P(flox.PnLTracker):
        def on_signal(self, sig):
            seen.append((sig.symbol, sig.side, sig.quantity))

    pnl = P()
    runner.set_pnl_tracker(pnl)
    runner.start()
    runner.on_trade(sym_id, 100.0, 1.0, True, 1_000)
    runner.stop()
    check(len(seen) == 1, f"PnLTracker.on_signal fired once, got {len(seen)}")
    check(seen[0][1] == "buy", f"side='buy', got {seen[0][1] if seen else 'N/A'}")


def test_storage_sink():
    print("test_storage_sink")
    reg, runner, strat, sym_id, _ = make_runner_with_strategy()
    seen = []

    class Sink(flox.StorageSink):
        def store(self, sig):
            seen.append(sig)

    runner.set_storage_sink(Sink())
    runner.start()
    runner.on_trade(sym_id, 100.0, 1.0, True, 1_000)
    runner.stop()
    check(len(seen) == 1, f"StorageSink.store fired once, got {len(seen)}")


def test_risk_manager_drops_signal():
    print("test_risk_manager_drops_signal")
    reg, runner, strat, sym_id, fired_signal = make_runner_with_strategy()

    class RM(flox.RiskManager):
        def allow(self, sig):
            return False  # always deny

    runner.set_risk_manager(RM())
    runner.start()
    runner.on_trade(sym_id, 100.0, 1.0, True, 1_000)
    runner.stop()
    check(len(fired_signal) == 0,
          f"on_signal callback should NOT fire when RM denies, got {len(fired_signal)}")


def test_kill_switch_drops_signal():
    print("test_kill_switch_drops_signal")
    reg, runner, strat, sym_id, fired_signal = make_runner_with_strategy()

    class KS(flox.KillSwitch):
        def check(self, sig):
            return False

    runner.set_kill_switch(KS())
    runner.start()
    runner.on_trade(sym_id, 100.0, 1.0, True, 1_000)
    runner.stop()
    check(len(fired_signal) == 0, "KillSwitch denies → no on_signal")


def test_order_validator_drops_signal():
    print("test_order_validator_drops_signal")
    reg, runner, strat, sym_id, fired_signal = make_runner_with_strategy()

    class V(flox.OrderValidator):
        def validate(self, sig):
            return False

    runner.set_order_validator(V())
    runner.start()
    runner.on_trade(sym_id, 100.0, 1.0, True, 1_000)
    runner.stop()
    check(len(fired_signal) == 0, "OrderValidator rejects → no on_signal")


def test_executor_receives_signals():
    print("test_executor_receives_signals")
    reg, runner, strat, sym_id, _ = make_runner_with_strategy()
    submits = []
    starts = [0]
    stops = [0]

    class E(flox.Executor):
        def submit(self, order):
            submits.append((order.symbol, order.side, order.quantity, order.order_type))

        def on_start(self):
            starts[0] += 1

        def on_stop(self):
            stops[0] += 1

    runner.set_executor(E())
    runner.start()
    check(starts[0] == 1, "Executor.on_start fired on runner.start()")
    runner.on_trade(sym_id, 100.0, 1.0, True, 1_000)
    runner.stop()
    check(stops[0] == 1, "Executor.on_stop fired on runner.stop()")
    check(len(submits) == 1, f"Executor.submit fired once, got {len(submits)}")
    if submits:
        check(submits[0][3] == "market", f"order_type='market', got {submits[0][3]}")


def test_market_data_recorder():
    print("test_market_data_recorder")
    reg = flox.SymbolRegistry()
    sym_id = reg.add_symbol("test", "BTC", 0.01)
    runner = flox.Runner(reg, lambda sig: None, threaded=False)

    class S(flox.Strategy):
        pass

    runner.add_strategy(S(symbols=[sym_id]))

    trades = []
    book_calls = [0]
    starts = [0]

    class R(flox.MarketDataRecorderHook):
        def on_trade(self, t):
            trades.append((t.symbol, t.price, t.is_buy))

        def on_book_update(self, symbol, is_snapshot, bids, asks, ts_ns):
            book_calls[0] += 1

        def on_start(self):
            starts[0] += 1

    runner.set_market_data_recorder(R())
    runner.start()
    check(starts[0] == 1, "MarketDataRecorderHook.on_start fired")
    runner.on_trade(sym_id, 101.5, 0.5, True, 5_000)
    runner.on_book_snapshot(sym_id, [100.0, 99.0], [1.0, 2.0],
                             [102.0, 103.0], [1.0, 2.0], 6_000)
    runner.stop()
    check(len(trades) == 1, f"on_trade fired once, got {len(trades)}")
    check(book_calls[0] == 1, "on_book_update fired once")


def test_execution_listener_with_backtest():
    print("test_execution_listener_with_backtest")
    import numpy as np

    reg = flox.SymbolRegistry()
    sym_id = reg.add_symbol("test", "BTC", 0.01)
    btr = flox.BacktestRunner(reg, fee_rate=0.0, initial_capital=10_000.0)

    fills = []
    submits = []

    class L(flox.ExecutionListener):
        def on_filled(self, order):
            fills.append((order.symbol, order.side, order.price, order.quantity))

        def on_submitted(self, order):
            submits.append(order)

    btr.add_execution_listener(L())

    fired = [False]

    class S(flox.Strategy):
        def on_bar(self, ctx, bar):
            if fired[0]:
                return
            fired[0] = True
            self.emit_market_buy(sym_id, 1.0)

    btr.set_strategy(S(symbols=[sym_id]))
    btr.run_bars(
        start_time_ns=np.array([1_000_000_000], dtype=np.int64),
        end_time_ns=np.array([1_999_999_999], dtype=np.int64),
        open=np.array([100.0]),
        high=np.array([101.0]),
        low=np.array([99.0]),
        close=np.array([100.5]),
        volume=np.array([10.0]),
        symbol="BTC",
    )
    check(len(fills) >= 1, f"ExecutionListener.on_filled fired ≥1×, got {len(fills)}")


def test_executor_with_backtest():
    print("test_executor_with_backtest")
    import numpy as np

    reg = flox.SymbolRegistry()
    sym_id = reg.add_symbol("test", "BTC", 0.01)
    btr = flox.BacktestRunner(reg, fee_rate=0.0, initial_capital=10_000.0)

    submits = []

    class E(flox.Executor):
        def submit(self, order):
            submits.append((order.symbol, order.side, order.quantity, order.order_type))

    btr.set_executor(E())

    fired = [False]

    class S(flox.Strategy):
        def on_bar(self, ctx, bar):
            if fired[0]:
                return
            fired[0] = True
            self.emit_market_buy(sym_id, 1.0)

    btr.set_strategy(S(symbols=[sym_id]))
    btr.run_bars(
        start_time_ns=np.array([1_000_000_000], dtype=np.int64),
        end_time_ns=np.array([1_999_999_999], dtype=np.int64),
        open=np.array([100.0]),
        high=np.array([101.0]),
        low=np.array([99.0]),
        close=np.array([100.5]),
        volume=np.array([10.0]),
        symbol="BTC",
    )
    check(len(submits) == 1, f"Custom Executor.submit fired once, got {len(submits)}")
    if submits:
        check(submits[0][3] == "market", f"order_type='market', got {submits[0][3]}")


def test_log_callback():
    print("test_log_callback")
    msgs = []

    def cb(level, msg):
        msgs.append((level, msg))

    flox.set_log_callback(cb)
    # Emit a log line — for now just detach to prove the wiring doesn't
    # crash. The C ABI lacks a public Python-driven log emitter.
    flox.set_log_callback(None)
    check(True, "set_log_callback wires + detaches without crash")


if __name__ == "__main__":
    test_pnl_tracker()
    test_storage_sink()
    test_risk_manager_drops_signal()
    test_kill_switch_drops_signal()
    test_order_validator_drops_signal()
    test_executor_receives_signals()
    test_market_data_recorder()
    test_execution_listener_with_backtest()
    test_executor_with_backtest()
    test_log_callback()
    print(f"\n{_passed} passed, {_failed} failed")
    sys.exit(0 if _failed == 0 else 1)

File: node/test/test_hooks.js
'use strict';
/**
 * node/test/test_hooks.js — smoke-test NAPI wrappers for the extension
 * hooks (PnLTracker, StorageSink, RiskManager, KillSwitch,
 * OrderValidator, MarketDataRecorderHook, Executor,
 * ExecutionListener, setLogCallback).
 *
 * Run from repo root:
 *   cd node && node test/test_hooks.js
 */

const path = require('path');
const flox = require(path.join(__dirname, '..'));

let passed = 0;
let failed = 0;
function check(condition, msg) {
  if (condition) {
    console.log(`  ok  ${msg}`);
    passed++;
  } else {
    console.error(`  FAIL  ${msg}`);
    failed++;
  }
}

// Helper — build a sync runner with a one-shot strategy that fires a
// market buy on the first onTrade.
function makeRunnerWithStrategy() {
  const reg = new flox.SymbolRegistry();
  const sym = reg.addSymbol('test', 'BTC', 0.01);
  const onSignalCalls = [];
  const runner = new flox.Runner(reg, (sig) => onSignalCalls.push(sig), false);

  let fired = false;
  runner.addStrategy({
    symbols: [sym],
    onTrade(_ctx, _trade, emit) {
      if (fired) return;
      fired = true;
      emit.marketBuy(Number(sym), 1.0);
    },
  });
  return { reg, runner, sym, onSignalCalls };
}

// ── PnLTracker ──────────────────────────────────────────────────────

function testPnlTracker() {
  console.log('testPnlTracker');
  const { runner, sym } = makeRunnerWithStrategy();
  const seen = [];
  runner.setPnlTracker({
    onSignal(sig) { seen.push({ sym: sig.symbol, side: sig.side, qty: sig.quantity }); },
  });
  runner.start();
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();
  check(seen.length === 1, `PnLTracker.onSignal fired once, got ${seen.length}`);
  check(seen[0]?.side === 'buy', `side='buy' (${seen[0]?.side})`);
}

// ── StorageSink ─────────────────────────────────────────────────────

function testStorageSink() {
  console.log('testStorageSink');
  const { runner, sym } = makeRunnerWithStrategy();
  const stored = [];
  runner.setStorageSink({ store(sig) { stored.push(sig); } });
  runner.start();
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();
  check(stored.length === 1, `StorageSink.store fired once, got ${stored.length}`);
}

// ── RiskManager ─────────────────────────────────────────────────────

function testRiskManagerDrops() {
  console.log('testRiskManagerDrops');
  const { runner, sym, onSignalCalls } = makeRunnerWithStrategy();
  runner.setRiskManager({ allow() { return false; } });
  runner.start();
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();
  check(onSignalCalls.length === 0,
        `RiskManager denies → no on_signal, got ${onSignalCalls.length}`);
}

// ── KillSwitch ──────────────────────────────────────────────────────

function testKillSwitchDrops() {
  console.log('testKillSwitchDrops');
  const { runner, sym, onSignalCalls } = makeRunnerWithStrategy();
  runner.setKillSwitch({ check() { return false; } });
  runner.start();
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();
  check(onSignalCalls.length === 0, `KillSwitch denies → no on_signal`);
}

// ── OrderValidator ──────────────────────────────────────────────────

function testOrderValidatorDrops() {
  console.log('testOrderValidatorDrops');
  const { runner, sym, onSignalCalls } = makeRunnerWithStrategy();
  runner.setOrderValidator({ validate() { return false; } });
  runner.start();
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();
  check(onSignalCalls.length === 0, `OrderValidator rejects → no on_signal`);
}

// ── Executor ────────────────────────────────────────────────────────

function testExecutor() {
  console.log('testExecutor');
  const { runner, sym } = makeRunnerWithStrategy();
  const submits = [];
  let starts = 0, stops = 0;
  runner.setExecutor({
    submit(o) { submits.push({ sym: o.symbol, side: o.side, qty: o.quantity, type: o.orderType }); },
    onStart() { starts++; },
    onStop() { stops++; },
  });
  runner.start();
  check(starts === 1, `Executor.onStart fired on runner.start (got ${starts})`);
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();
  check(stops === 1, `Executor.onStop fired on runner.stop (got ${stops})`);
  check(submits.length === 1, `Executor.submit fired once (got ${submits.length})`);
  check(submits[0]?.type === 'market', `submit.orderType='market' (${submits[0]?.type})`);
}

// ── MarketDataRecorderHook ──────────────────────────────────────────

function testMarketDataRecorder() {
  console.log('testMarketDataRecorder');
  const reg = new flox.SymbolRegistry();
  const sym = reg.addSymbol('test', 'BTC', 0.01);
  const runner = new flox.Runner(reg, () => {}, false);
  runner.addStrategy({ symbols: [sym] });

  const trades = [];
  const books = [];
  let starts = 0;
  runner.setMarketDataRecorder({
    onTrade(t) { trades.push(t); },
    onBookUpdate(s, snap, bids, asks) { books.push({ s, snap, bids, asks }); },
    onStart() { starts++; },
  });
  runner.start();
  check(starts === 1, `MarketDataRecorderHook.onStart fired (${starts})`);
  runner.onTrade(Number(sym), 101.5, 0.5, true, 5000);
  // index.d.ts declares Float64Array for these four arguments.
  runner.onBookSnapshot(Number(sym),
    new Float64Array([100, 99]), new Float64Array([1, 2]),
    new Float64Array([102, 103]), new Float64Array([1, 2]),
    6000);
  runner.stop();
  check(trades.length === 1, `recorder.onTrade fired once (got ${trades.length})`);
  check(books.length === 1, `recorder.onBookUpdate fired once`);
  // bids/asks are flat BigInt64Array views: [price_raw, qty_raw, ...].
  // 2 levels × 2 int64s per level = 4 elements.
  check(books[0]?.bids instanceof BigInt64Array, `bids is BigInt64Array`);
  check(books[0]?.bids?.length === 4, `bids has 4 int64s (2 levels)`);
}

// ── ExecutionListener (BacktestRunner) ──────────────────────────────

function testExecutionListenerBacktest() {
  console.log('testExecutionListenerBacktest');
  const reg = new flox.SymbolRegistry();
  const sym = reg.addSymbol('test', 'BTC', 0.01);
  const btr = new flox.BacktestRunner(reg, 0.0, 10000.0);

  const fills = [];
  btr.addExecutionListener({
    onFilled(o) { fills.push(o); },
  });

  let fired = false;
  btr.setStrategy({
    symbols: [sym],
    onBar(_ctx, _bar, emit) {
      if (fired) return;
      fired = true;
      emit.marketBuy(Number(sym), 1.0);
    },
  });

  btr.runBars(
    new BigInt64Array([1_000_000_000n]),
    new BigInt64Array([1_999_999_999n]),
    new Float64Array([100.0]),
    new Float64Array([101.0]),
    new Float64Array([99.0]),
    new Float64Array([100.5]),
    new Float64Array([10.0]),
    'BTC');

  // Fill counts may vary by simulator quirks — assert at least one fill
  // observed via the listener.
  check(fills.length >= 1, `ExecutionListener.onFilled ≥1× (got ${fills.length})`);
}

// ── Executor (BacktestRunner) ────────────────────────────────────────

function testExecutorBacktest() {
  console.log('testExecutorBacktest');
  const reg = new flox.SymbolRegistry();
  const sym = reg.addSymbol('test', 'BTC', 0.01);
  const btr = new flox.BacktestRunner(reg, 0.0, 10000.0);

  const submits = [];
  btr.setExecutor({ submit(o) { submits.push(o); } });

  let fired = false;
  btr.setStrategy({
    symbols: [sym],
    onBar(_ctx, _bar, emit) {
      if (fired) return;
      fired = true;
      emit.marketBuy(Number(sym), 1.0);
    },
  });

  btr.runBars(
    new BigInt64Array([1_000_000_000n]),
    new BigInt64Array([1_999_999_999n]),
    new Float64Array([100.0]),
    new Float64Array([101.0]),
    new Float64Array([99.0]),
    new Float64Array([100.5]),
    new Float64Array([10.0]),
    'BTC');

  check(submits.length === 1, `Custom Executor.submit fired once (got ${submits.length})`);
  check(submits[0]?.orderType === 'market', `orderType='market'`);
}

// ── LP signal fields ────────────────────────────────────────────────

function testLpSignalFields() {
  console.log('testLpSignalFields');
  const reg = new flox.SymbolRegistry();
  const sym = reg.addSymbol('test', 'ETH-USDC', 0.01);
  const onSignalCalls = [];
  const runner = new flox.Runner(reg, (sig) => onSignalCalls.push(sig), false);

  let fired = false;
  runner.addStrategy({
    symbols: [sym],
    onTrade(_ctx, _trade, emit) {
      if (fired) return;
      fired = true;
      // provideLiquidity(priceLower, priceUpper, liquidity, [symbol]) and
      // withdrawLiquidity(liquidity, [symbol]) -- symbol trails and is
      // optional, same convention as marketBuy/marketSell/etc.
      emit.provideLiquidity(1800.0, 2200.0, 5.0, Number(sym));
      emit.withdrawLiquidity(2.5, Number(sym));
    },
  });
  runner.start();
  runner.onTrade(Number(sym), 100, 1, true, 1000);
  runner.stop();

  check(onSignalCalls.length === 2, `both liquidity signals fired, got ${onSignalCalls.length}`);
  const [provide, withdraw] = onSignalCalls;
  check(provide?.orderType === 'provide_liquidity',
        `first signal is provide_liquidity, got ${provide?.orderType}`);
  check(Math.abs(provide?.rangeLower - 1800.0) < 1e-9,
        `rangeLower crosses the boundary, got ${provide?.rangeLower}`);
  check(Math.abs(provide?.rangeUpper - 2200.0) < 1e-9,
        `rangeUpper crosses the boundary, got ${provide?.rangeUpper}`);
  check(Math.abs(provide?.liquidity - 5.0) < 1e-9,
        `liquidity crosses the boundary, got ${provide?.liquidity}`);
  check(withdraw?.orderType === 'withdraw_liquidity',
        `second signal is withdraw_liquidity, got ${withdraw?.orderType}`);
  check(Math.abs(withdraw?.liquidity - 2.5) < 1e-9,
        `withdraw liquidity crosses the boundary, got ${withdraw?.liquidity}`);
}

// ── setLogCallback ──────────────────────────────────────────────────

function testLogCallback() {
  console.log('testLogCallback');
  // Just exercise the wiring without expecting log output (the C ABI
  // doesn't have a public Node-driven log emitter for testing).
  flox.setLogCallback((_lvl, _msg) => {});
  flox.setLogCallback(null);
  check(true, `setLogCallback wires + detaches without crash`);
}

// ── Run ─────────────────────────────────────────────────────────────

testPnlTracker();
testStorageSink();
testRiskManagerDrops();
testKillSwitchDrops();
testOrderValidatorDrops();
testExecutor();
testMarketDataRecorder();
testExecutionListenerBacktest();
testExecutorBacktest();
testLpSignalFields();
testLogCallback();

console.log(`\n${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);
